### PR TITLE
Speed up `find_files_and_readers`

### PR DESCRIPTION
**Pull request by mraspaud**
This PR is an attempt to speed up the `find_files_and_readers` function.

 - [ ] Tests added 
 - [ ] Fully documented

### CODE REVIEW

**Comment by mraspaud** [comment]
The first action is to replace yaml's `UnsafeLoader` with `CLoader`. The latter makes use of C's `libyaml` and cuts down a significant amount of time, from 57s to 7s for the yaml reading alone in the TestFindFilesAndReaders run.

As a note, I couldn't find a reference to `UnsafeLoader` in the pyyaml documentation, so I suspect `Loader` is what we should use.

**Comment by mraspaud** [comment]
Globally over these tests, that change alone made the tests run more than twice faster. The big item now (half of the time) is going through the entry points. The `importlib.metadata.entry_points()` call/iteration takes now half of the time. @djhoese you looked at the entry_points problem, right?

**Comment by djhoese** [comment]
Interesting on the CLoader stuff. Make sure you search for all yaml imports. There might be others.

Which `entry_points` usage are you looking at/having trouble with? The plugins stuff? In general, at least in the past, entry_points can/could be slow if there were a lot of packages installed...or at least `pkg_resources` had that problem. I thought newer versions did a better job of that kind of stuff.

**Comment by mraspaud** [comment]
But I do have a lot of packages installed

**Comment by djhoese** [comment]
Semi-related, but mostly for scripts not plugins, I just found:

https://github.com/ninjaaron/fast-entry_points

**Comment by djhoese** [comment]
I guess the question is, what part of the entry point lookup takes long? Is it the imports or the actual sorting/searching for the entry points? Is it finding any that are installed?

**Comment by mraspaud** [comment]
image
Here is the snakeviz for the entry_points call in one test.
So out of 4 seconds this takes, 2 seconds seems to be for IO, the rest for parsing and sorting. Not sure there is much we can do about this, appart from possibly caching the entry points to avoid multiple calls? Although I'm not sure that would work for pytest.

**Comment by mraspaud** [comment]
So, turned out the `entry_points` function was called many times, because every time a yaml config file was loaded, we looked for entry points and called that function. By making a small wrapper around it for memoization, I got a 50% improvement in time overall the `find_file_and_readers` test. 

The tests now run 5 times faster than in the main branch, I'm pretty happy about it. So I'm marking this as ready for review.

**Comment by mraspaud** [comment]
So overall, the only loss in functionality we have with this is that entry points won't be rechecked every time we load a config file, thus we have to restart python or reimport satpy to get the effects of a new plugin. I think this is acceptable.

File: satpy/_config.py [hunk 1]
```diff
@@ -18,6 +18,11 @@
 """Satpy Configuration directory and file handling."""
 from __future__ import annotations
 
+try:
+    from functools import cache  # type: ignore
+except ImportError:  # for python < 3.9
+    from functools import lru_cache as cache
```
Comment (djhoese) [inline satpy/_config.py:24]: You could import this from the compat module:

https://github.com/ameraner/satpy/blob/d85fc1843aeb2fa2f8cf4dd5d395ffa7580e5083/satpy/_compat.py#L93-L96

**Review by djhoese — APPROVED** [review]
Looks good to me. I was wondering if the entire `get_entry_points_config_dirs` could be cached, but since it depends on `config.get('config_path')` we'd have to include that in the cache hash which makes the code uglier for very little gain I think. That said, I have a decorator like that in the composite loading.

File: satpy/_config.py (hunk 1 above)
Comment (mraspaud) [reply satpy/_config.py:24]: ah, yes, of course! thank you

**Comment by djhoese** [comment]
Ah the plugin tests are failing. You should clear the cache in this function:

https://github.com/pytroll/satpy/blob/bf5fec47ebb2e012843041bec5a45128285790b4/satpy/tests/conftest.py#L43-L47

Or I guess do the same type of logic but only in the plugin tests?

**Comment by mraspaud** [comment]
strange, the tests pass locally...

**Comment by mraspaud** [comment]
my bad, they do not pass locally

**Comment by mraspaud** [comment]
Ok, I'm clearing the cache now in the plugin tests. Let's hope this fixed the CI.

**Pull request merged**